File: src/java/delta/Diff.java
package delta;

import java.util.ArrayList;
import java.util.List;

import static delta.Types.*;

/** Dispatch to the appropriate differencing algorithm. */
public final class Diff {
    private Diff() {}

    // ── Shared utilities used by all three algorithm implementations ──

    /** Compare byte subarrays for equality. */
    static boolean arrayEquals(byte[] a, int aOff, byte[] b, int bOff, int len) {
        for (int i = 0; i < len; i++) {
            if (a[aOff + i] != b[bOff + i]) return false;
        }
        return true;
    }

    /** Copy a range of bytes into a new array. */
    static byte[] copyRange(byte[] data, int from, int to) {
        byte[] result = new byte[to - from];
        System.arraycopy(data, from, result, 0, to - from);
        return result;
    }

    /** Print delta compression statistics to stderr. */
    static void printStats(List<Command> commands) {
        List<Integer> copyLens = new ArrayList<>();
        long totalCopy = 0, totalAdd = 0;
        int numCopies = 0, numAdds = 0;
        for (Command cmd : commands) {
            if (cmd instanceof CopyCmd) {
                CopyCmd c = (CopyCmd) cmd;
                totalCopy += c.length;
                numCopies++;
                copyLens.add(c.length);
            } else if (cmd instanceof AddCmd) {
                AddCmd a = (AddCmd) cmd;
                totalAdd += a.data.length;
                numAdds++;
            }
        }
        long totalOut = totalCopy + totalAdd;
        double copyPct = totalOut > 0 ? totalCopy * 100.0 / totalOut : 0;
        System.err.printf("  result: %d copies (%d bytes), %d adds (%d bytes)%n" +
            "  result: copy coverage %.1f%%, output %d bytes%n",
            numCopies, totalCopy, numAdds, totalAdd, copyPct, totalOut);
        if (!copyLens.isEmpty()) {
            copyLens.sort(null);
            double mean = totalCopy / (double) copyLens.size();
            int median = copyLens.get(copyLens.size() / 2);
            System.err.printf("  copies: %d regions, min=%d max=%d mean=%.1f median=%d bytes%n",
                copyLens.size(), copyLens.get(0), copyLens.get(copyLens.size() - 1), mean, median);
        }
    }

    public static List<Command> diff(Algorithm algo, byte[] r, byte[] v,
                                     DiffOptions opts) {
        switch (algo) {
            case GREEDY:
                return Greedy.diff(r, v, opts);
            case ONEPASS:
                return Onepass.diff(r, v, opts);
            case CORRECTING:
                return Correcting.diff(r, v, opts);
            default:
                throw new IllegalArgumentException("unknown algorithm: " + algo);
        }
    }

    public static List<Command> diffDefault(Algorithm algo, byte[] r, byte[] v) {
        return diff(algo, r, v, new DiffOptions());
    }
}

File: src/go/delta/diff.go
package delta

import (
	"fmt"
	"os"
	"sort"
)

// ── Shared utilities used by all three algorithm implementations ──

// regionEquals compares len bytes of a[aOff:] and b[bOff:].
func regionEquals(a []byte, aOff int, b []byte, bOff int, length int) bool {
	for i := 0; i < length; i++ {
		if a[aOff+i] != b[bOff+i] {
			return false
		}
	}
	return true
}

// printStats writes compression statistics to stderr.
func printStats(commands []Command) {
	var copyLens []int
	var totalCopy, totalAdd int64
	numCopies, numAdds := 0, 0
	for _, cmd := range commands {
		switch c := cmd.(type) {
		case CopyCmd:
			totalCopy += int64(c.Length)
			numCopies++
			copyLens = append(copyLens, c.Length)
		case AddCmd:
			totalAdd += int64(len(c.Data))
			numAdds++
		}
	}
	totalOut := totalCopy + totalAdd
	var copyPct float64
	if totalOut > 0 {
		copyPct = float64(totalCopy) * 100.0 / float64(totalOut)
	}
	fmt.Fprintf(os.Stderr, "  result: %d copies (%d bytes), %d adds (%d bytes)\n"+
		"  result: copy coverage %.1f%%, output %d bytes\n",
		numCopies, totalCopy, numAdds, totalAdd, copyPct, totalOut)
	if len(copyLens) > 0 {
		sort.Ints(copyLens)
		mean := float64(totalCopy) / float64(len(copyLens))
		median := copyLens[len(copyLens)/2]
		fmt.Fprintf(os.Stderr, "  copies: %d regions, min=%d max=%d mean=%.1f median=%d bytes\n",
			len(copyLens), copyLens[0], copyLens[len(copyLens)-1], mean, median)
	}
}

// DiffErr dispatches to the requested differencing algorithm.
func DiffErr(algo Algorithm, r, v []byte, opts DiffOptions) ([]Command, error) {
	switch algo {
	case AlgorithmGreedy:
		return diffGreedy(r, v, opts), nil
	case AlgorithmOnepass:
		return diffOnepass(r, v, opts), nil
	case AlgorithmCorrecting:
		return diffCorrecting(r, v, opts), nil
	default:
		return nil, fmt.Errorf("unknown algorithm: %d", int(algo))
	}
}

// Diff dispatches to the requested differencing algorithm.
//
// It returns nil for an unknown algorithm; callers that need explicit error
// handling should use DiffErr.
func Diff(algo Algorithm, r, v []byte, opts DiffOptions) []Command {
	cmds, err := DiffErr(algo, r, v, opts)
	if err != nil {
		return nil
	}
	return cmds
}

// DiffDefaultErr runs the given algorithm with default options.
func DiffDefaultErr(algo Algorithm, r, v []byte) ([]Command, error) {
	return DiffErr(algo, r, v, DefaultDiffOptions())
}

// DiffDefault runs the given algorithm with default options.
func DiffDefault(algo Algorithm, r, v []byte) []Command {
	return Diff(algo, r, v, DefaultDiffOptions())
}
